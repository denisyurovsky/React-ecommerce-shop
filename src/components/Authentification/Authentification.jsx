import React from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { authStatus } from '../../helpers/constants/authConstants';
import {
  getLoginState,
  getRegisterState,
  loginUser,
  registerUser,
  removeLockAfterTimeout,
  resetError,
} from '../../store/user/userSlice';
import { ProfileMenu } from '../ProfileMenu/ProfileMenu';

import AuthModal from './AuthModal/AuthModal';

const { FULFILLED, LOCKED } = authStatus;

const Authentification = () => {
  const dispatch = useDispatch();
  const loginStatus = useSelector(getLoginState);
  const registerStatus = useSelector(getRegisterState);
  const loginMessage = useSelector((state) => state.user.loginError);
  const registerMessage = useSelector((state) => state.user.registerError);

  const isLogged = loginStatus === FULFILLED || registerStatus === FULFILLED;

  const sendLoginForm = ({ password, email }) => {
    dispatch(loginUser({ password, email }));
  };

  const sendRegisterForm = ({ email, password, ...info }) => {
    dispatch(registerUser({ email, password, ...info }));
  };

  if (loginStatus === LOCKED) {
    dispatch(removeLockAfterTimeout());
  }

  const resetState = () => dispatch(resetError());

  return (
    <>
      {isLogged ? (
        <ProfileMenu />
      ) : (
        <AuthModal
          registerUser={sendRegisterForm}
          registerStatus={registerStatus}
          loginUser={sendLoginForm}
          loginStatus={loginStatus}
          loginErrorMessage={loginMessage}
          registerErrorMessage={registerMessage}
          resetError={resetState}
        />
      )}
    </>
  );
};

export default Authentification;
