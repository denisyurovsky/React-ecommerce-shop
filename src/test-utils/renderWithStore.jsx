import { configureStore } from '@reduxjs/toolkit';
import { render as rtlRender } from '@testing-library/react';
import PropTypes from 'prop-types';
import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import cartReducer from '../store/cart/cartSlice';
import categoriesReducer from '../store/categories/categoriesSlice';
import feedbackReducer from '../store/feedback/feedbackSlice';
import productsReducer from '../store/products/productsSlice';
import userReducer from '../store/user/userSlice';

function render(component, { role = null, store } = {}) {
  let preloadedState;

  if (!store) {
    preloadedState = role
      ? {
          user: {
            user: {
              id: 5,
              role,
            },
          },
        }
      : {};
  }

  const storeWithState = store
    ? store
    : configureStore({
        reducer: {
          cart: cartReducer,
          products: productsReducer,
          categories: categoriesReducer,
          user: userReducer,
          feedback: feedbackReducer,
        },
        preloadedState,
      });

  function Wrapper({ children }) {
    return (
      <Provider store={storeWithState}>
        {children}
        <ToastContainer />
      </Provider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.object,
  };

  return rtlRender(component, { wrapper: Wrapper });
}

export * from '@testing-library/react';
export default render;
