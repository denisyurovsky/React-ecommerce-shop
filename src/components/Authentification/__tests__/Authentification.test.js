import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import React from 'react';

import {
  MAX_LOGIN_ATTEMPTS,
  ERROR,
} from '../../../helpers/constants/authConstants';
import renderWithStore, { screen } from '../../../test-utils/renderWithStore';
import Authentification from '../Authentification';

const successfulHandlers = [
  rest.post('/login', (req, res, ctx) => {
    if (
      req.body.email !== 'ivan@yandex.ru' &&
      req.body.password !== 'password'
    ) {
      return res(
        ctx.status(400),
        ctx.json({ message: 'Request failed with status code 400' })
      );
    }

    return res(
      ctx.json({
        accessToken: 'mocked_user_token_login',
        user: {
          id: 1,
        },
      }),
      ctx.delay(150)
    );
  }),
  rest.post('/register', (req, res, ctx) => {
    if (
      req.body.firstName !== 'Ivan' &&
      req.body.lastName !== 'Ivn' &&
      req.body.email !== 'ivan@yandex.ru' &&
      req.body.password !== 'AbcDefG12!'
    ) {
      res(
        ctx.status(400),
        ctx.json({ message: 'Request failed with status code 400' })
      );
    }

    return res(
      ctx.json({
        accessToken: 'mocked_user_token_register',
        user: {
          id: 1,
        },
      }),
      ctx.delay(150)
    );
  }),
];

const failedHandlers = [
  rest.post('/login', (req, res, ctx) => {
    return res(
      ctx.status(400),
      ctx.json({ message: 'Request failed with status code 400' })
    );
  }),
  rest.post('/register', (req, res, ctx) => {
    return res(
      ctx.status(400),
      ctx.json({ message: 'Request failed with status code 400' })
    );
  }),
];

const sendLoginRequest = () => {
  userEvent.click(screen.getByTestId('btn-login'));
  fillInLoginForms();
  userEvent.click(screen.getByRole('button', { name: /sign in/i }));
};

const fillInLoginForms = () => {
  userEvent.type(screen.getByLabelText('Email'), 'ivan@yandex.ru');
  userEvent.type(screen.getByLabelText('Password'), 'password');
};

const sendRegisterRequest = () => {
  userEvent.click(screen.getByTestId('btn-login'));
  userEvent.click(screen.getByRole('tab', { name: /sign up/i }));
  fillInRegisterForms();
  userEvent.click(screen.getByRole('button', { name: /sign up/i }));
};

const fillInRegisterForms = () => {
  userEvent.type(screen.getByLabelText('First name'), 'Ivan');
  userEvent.type(screen.getByLabelText('Last name'), 'Ivn');
  userEvent.type(screen.getByLabelText('Email'), 'ivan@yandex.ru');
  userEvent.type(screen.getByLabelText('Password'), 'AbcDefG12!');
  userEvent.type(screen.getByLabelText('Confirm password'), 'AbcDefG12!');
};

describe('Successfull login and register', () => {
  const server = setupServer(...successfulHandlers);

  beforeAll(() => server.listen());
  afterAll(() => server.close());
  beforeEach(() => {
    renderWithStore(<Authentification />);
    localStorage.clear();
  });

  it('should allow user to login', async () => {
    sendLoginRequest();

    await screen.findByTestId('btn-profile');

    expect(localStorage.getItem('accessToken')).toBe('mocked_user_token_login');
  });

  it('should allow user to register', async () => {
    sendRegisterRequest();

    await screen.findByTestId('btn-profile');

    expect(localStorage.getItem('accessToken')).toEqual(
      'mocked_user_token_register'
    );
  });
});

describe('Failed server:', () => {
  const server = setupServer(...failedHandlers);

  beforeAll(() => server.listen());
  afterAll(() => server.close());

  describe('Failed login and register:', () => {
    beforeEach(() => {
      renderWithStore(<Authentification />);
      localStorage.clear();
    });

    it('should not allow user to login with wrong data', async () => {
      sendLoginRequest();

      expect(await screen.findAllByText(ERROR.LOGIN)).toHaveLength(2);
      expect(localStorage.getItem('accessToken')).toBeNull();
    });

    it('should remove server error after switches between tabs', async () => {
      sendLoginRequest();

      await screen.findAllByText(ERROR.LOGIN);
      userEvent.click(screen.getByRole('tab', { name: /sign up/i }));
      userEvent.click(screen.getByRole('tab', { name: /sign in/i }));

      expect(screen.queryByText(ERROR.LOGIN)).toBeNull();
    });

    it('should be able to detect incorrect data in sign up', async () => {
      sendRegisterRequest();

      expect(await screen.findByText(ERROR.REGISTER)).toBeInTheDocument();
      expect(localStorage.getItem('accessToken')).toBeNull();
    });
  });

  describe('Locked login state:', () => {
    const createLockedState = async () => {
      sendLoginRequest();

      for (let i = 0; i < MAX_LOGIN_ATTEMPTS - 1; i++) {
        await screen.findAllByText(ERROR.LOGIN);
        userEvent.click(screen.getByRole('button', { name: /sign in/i }));
      }
    };

    beforeAll(() => jest.useFakeTimers());
    beforeEach(() => {
      renderWithStore(<Authentification />);
      localStorage.clear();
    });

    it('should display error', async () => {
      await createLockedState();
      await screen.findAllByText(ERROR.LOCK);

      expect(localStorage.getItem('accessToken')).toBeNull();
      expect(screen.getByRole('presentation')).toMatchSnapshot();
    });

    it('should display locked state after switches between tabs', async () => {
      await createLockedState();
      await screen.findAllByText(ERROR.LOCK);

      userEvent.click(screen.getByRole('tab', { name: /sign up/i }));
      userEvent.click(screen.getByRole('tab', { name: /sign in/i }));

      expect(screen.queryAllByText(ERROR.LOCK)).toHaveLength(2);
    });

    it('should unlock forms after delay', async () => {
      await createLockedState();
      await screen.findAllByText(ERROR.LOCK);

      jest.runOnlyPendingTimers();
      expect(screen.queryAllByText(ERROR.LOCK)).toHaveLength(0);
      expect(screen.getByRole('presentation')).toMatchSnapshot();
    });
  });
});
