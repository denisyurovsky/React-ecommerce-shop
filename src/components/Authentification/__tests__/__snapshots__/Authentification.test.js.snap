// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Failed server: Locked login state: should display error 1`] = `
<div
  class="MuiModal-root container css-79ws1d-MuiModal-root"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="modal MuiBox-root css-0"
    tabindex="-1"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall cross css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiTabs-root navigation css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer MuiTabs-centered css-1wf8b0h-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1pec19d-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Sign in
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1pec19d-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Sign up
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      class="content MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-disabled Mui-error MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-36"
            id="mui-36-label"
          >
            Email
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-disabled Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-describedby="mui-36-helper-text"
              aria-invalid="true"
              autocomplete="off"
              class="MuiInput-input MuiInputBase-input Mui-disabled css-1x51dt5-MuiInputBase-input-MuiInput-input"
              disabled=""
              id="mui-36"
              type="text"
              value="ivan@yandex.ru"
            />
          </div>
          <p
            class="MuiFormHelperText-root Mui-disabled Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
            id="mui-36-helper-text"
          >
            Many wrong attempts, try again later
          </p>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary Mui-disabled Mui-error MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-37"
            id="mui-37-label"
          >
            Password
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary Mui-disabled Mui-error MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-ghsjzk-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-describedby="mui-37-helper-text"
              aria-invalid="true"
              autocomplete="off"
              class="MuiInput-input MuiInputBase-input Mui-disabled MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
              disabled=""
              id="mui-37"
              type="password"
              value="password"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <span
                aria-disabled="true"
                class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked Mui-disabled css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                data-testid="visibility"
                tabindex="-1"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  disabled=""
                  type="checkbox"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityOffIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <p
            class="MuiFormHelperText-root Mui-disabled Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-filled css-1d1r5q-MuiFormHelperText-root"
            id="mui-37-helper-text"
          >
            Many wrong attempts, try again later
          </p>
        </div>
        <label
          class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd css-1apuumd-MuiFormControlLabel-root"
        >
          <span
            aria-disabled="true"
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root Mui-disabled MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked Mui-disabled css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            tabindex="-1"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              disabled=""
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-ahj2mt-MuiTypography-root"
          >
            Remember me
          </span>
        </label>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root Mui-disabled  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          Sign in
        </button>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;

exports[`Failed server: Locked login state: should unlock forms after delay 1`] = `
<div
  class="MuiModal-root container css-79ws1d-MuiModal-root"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-test="sentinelStart"
    tabindex="0"
  />
  <div
    class="modal MuiBox-root css-0"
    tabindex="-1"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall cross css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      class="MuiTabs-root navigation css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer MuiTabs-centered css-1wf8b0h-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1pec19d-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            Sign in
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1pec19d-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            Sign up
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      class="content MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-57"
            id="mui-57-label"
          >
            Email
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-ghsjzk-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-57"
              type="text"
              value="ivan@yandex.ru"
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-58"
            id="mui-58-label"
          >
            Password
          </label>
          <div
            class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-ghsjzk-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              autocomplete="off"
              class="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
              id="mui-58"
              type="password"
              value="password"
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-standard MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <span
                class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                data-testid="visibility"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  type="checkbox"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityOffIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78 3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
            </div>
          </div>
        </div>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1apuumd-MuiFormControlLabel-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Remember me
          </span>
        </label>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Sign in
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    data-test="sentinelEnd"
    tabindex="0"
  />
</div>
`;
